Prevent duplicated bundle from including in block

NingLin-P opened the pull request:
close #1056 

This PR prevents the duplicated bundle from including in block by tracking the bundles of the last K blocks on the client side and using that to check duplicated bundles in the transaction pool.

More specifically, this PR introduces a new worker to maintain a queue `bundle_stored_in_last_k`, containing the bundles of the last K blocks (only contain the hash of the bundle currently), sorted from the bundle of newer block to the bundle of older block: 

**Initialization**
when a node is started and the first best block with block number N is imported, it travels back the canonical chain for blocks with block number from  N to N - K, finds all the bundle within each block, and pushes these bundles to `bundle_stored_in_last_k`.

**Update on block import**
whenever a new best block is imported, `bundle_stored_in_last_k` is updated as follows: 

```rust
let route = sp_blockchain::tree_route(&self.client, curent_best_hash, imcoming_block_hash)?;
let (retracted, enacted) = (route.retracted(), route.enacted());

// Remove the bundle of block from the stale fork
for retracted_block in retracted {
    // Remove the bundle from the head of `bundle_stored_in_last_k`
}

// Remove the bundle of block from the best fork
for enacted_block in retracted {
    // Add the bundle to the head of `bundle_stored_in_last_k`
}

// Remove bundle from the tail of `bundle_stored_in_last_k` to keep at most the bundle of the last K blocks
bundle_stored_in_last_k..truncate(confirm_depth_k);
```

**Check duplicate bundle**
When a bundle is submitted to the txn pool, perform the check as follows:
```rust
for block in bundle_stored_in_last_k.iter() {
    if block.bundle_hashs.contains(&incoming_bundle) {
        return Err(..);
    }
}
```


### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

liuchengxu commented on crates/subspace-bundle-checker/Cargo.toml:
```diff
@@ -0,0 +1,25 @@
+[package]
+name = "subspace-bundle-checker"
```
Something like `subspace-tx-pool-bundle-validator` feels more informative, and then `BundleChecker` can be `BundleValidator`.

NingLin-P replied:
Although the check is performed at txn pool, within this crate there is nothing related to txn pool, `subspace-tx-pool-bundle-validator` could be confusing, how about `subspace-bundle-validator`?

liuchengxu replied:
`subspace-bundle-validator` is not concrete/accurate as the bundle validation takes place in both the tx pool and runtime, `subspace-bundle-validator` makes me feel that all the bundle validations are covered in this place, but it's not in fact. Why `subspace-tx-pool-bundle-validator` is confusing? It can be even part of `subspace-transaction-pool` as a module.

NingLin-P replied:
> Why subspace-tx-pool-bundle-validator is confusing?

As a standalone crate, its name should reflect its functionality instead of where it is used outside.

> It can be even part of subspace-transaction-pool as a module.

Well, this is a better option compared to `subspace-tx-pool-bundle-validator` I think.

liuchengxu replied:
Well, `subspace-tx-pool-bundle-validator` exactly tells the function it serves, which is the bundle validator inside the tx pool and this is everything this crate provides.

NingLin-P replied:
Bundle validation is the function this crate provides, and where to use it whether in txn pool, block import, or any other place is a decision made out of this crate. If we want to emphasize that it should ever only be used in txn pool then making it a part of `subspace-transaction-pool` as a module is indeed a better option I think. But I will change it to `subspace-tx-pool-bundle-validator` if you insist.

liuchengxu replied:
>it should ever only be used in txn pool

This is true, the bundle validation logic this crate is scoped to the tx pool only and not intended to be used in any other places (at least for now).

Either it's a seperate crate or a module of `subspace-transaction-pool` is fine to me, but if you make it a crate, I'd expect it's named in a non-confusing way.

liuchengxu commented on crates/subspace-bundle-checker/src/lib.rs:
```diff
@@ -0,0 +1,295 @@
+    // will be recovered after successfully handling the next best block.
+    fn check_duplicate_bundle(
+        &self,
+        bundle: &Bundle<OpaqueExtrinsic, NumberFor<Block>, Block::Hash, DomainHash>,
```
Since only `bundle_hash` matters, passing `bundle_hash` directly would be clearer.

liuchengxu commented on crates/subspace-bundle-checker/src/lib.rs:
```diff
@@ -0,0 +1,295 @@
+    // The block that containing the bundles
+    block_hash: Block::Hash,
+    // The hashs of all the bundles within a block
+    bundle_hashs: HashSet<Hash>,
```
`hashs` => `hashes`. Also, I personally would probably use `VecDeque<(Block::Hash, HashSet<Hash>)>` below for simplicity as this abstraction somehow does not look satisfying.

NingLin-P replied:
Okay, I don't have a strong opinion about this struct, and `(Block::Hash, HashSet<Hash>)` is also fine to me.

liuchengxu commented on crates/subspace-transaction-pool/src/lib.rs:
```diff
@@ -169,8 +152,15 @@ where
+                        }
+                        if let Err(err) =
+                            self.bundle_checker.validate_receipts(*at, bundle.receipts)
+                        {
```
If all the validation logics are performed inside the validator, this can be simplified as `self.bundle_validator.validate_bundle(&bundle)`.

NingLin-P replied:
Yes, this idea also has come across my mind but I forget it somehow...

liuchengxu commented on crates/subspace-bundle-checker/src/lib.rs:
```diff
@@ -0,0 +1,295 @@
+        at: BlockId<Block>,
+        receipts: Vec<ExecutionReceipt<NumberFor<Block>, Block::Hash, DomainHash>>,
+    ) -> Result<(), sp_blockchain::Error>;
+}
```
The receipts are part of the bundle, it's ergonomic to provide such an API in which all the validation logic is performed within this crate.

```rust
pub trait ValidateBundle<Block: BlockT> {
    // for consensus chain, check the duplicated bundle and receipts.
    // for system domain, checks nothing.
    fn validate_bundle(&self, bundle: &Bundle) -> Result<(), BundleError>;
}
```

There is a difference between the consensus chain and system domain: the consensus chain checks the duplicated bundle and receipts but the system domain checks nothing on a bundle (the system domain only verifies the fraud proof in the tx pool), `()` as an empty impl is used below for the system domain which is not good regarding the readability, nice to make `()` something more explicit like a struct `SkipBundleValidation` that it performs no checks on the bundle.

NingLin-P replied:
`SkipBundleValidation` is indeed more informative!

liuchengxu commented on crates/subspace-bundle-checker/src/lib.rs:
```diff
@@ -0,0 +1,295 @@
+
+/// `BundleSyncer` used to share the bundles of the last K blocks with different thread
+#[derive(Clone)]
+struct BundleSyncer<Block: BlockT> {
```
`BundleSyncer` is sort of vague and hard to understand what it's for, naming is hard as always :P Just my two cents: `BundleStoredInLastK`/`RecentStoredBundlesMaintainer`.

liuchengxu commented on crates/subspace-service/src/lib.rs:
```diff
@@ -503,6 +522,12 @@ where
+        "bundle-collector",
+        None,
+        Box::pin(bundle_collector.run()),
+    );
```
This doesn't seem to be an essential task.

It's hard to read what this task is for because it's invisible in the architecture and the name is too vague. In addition, the initialization of stored bundles in the last k is implicit on the first imported block notification, which is not intuitive and incorrect even because new bundles can be received before the very first block notification.

1. Initialize the bundles stored in the last K on start up.
2. Update the state of recent bundles on each new block.

```rust
    task_manager.spaw_handle().spawn(
        "maintain-bundles-stored-in-last-k",
        None,
        Box::pin({
            initialize_recent_stored_bundles();

            client.import_notification_stream()
                .for_each(|notification| {
                 if notification.is_new_best {
                       update_recent_stored_bundles();
                 }
             })
        }),
    );
```

NingLin-P replied:
> It's hard to read what this task is for because it's invisible in the architecture and the name is too vague.

The naming could definitely be improved, but I am not sure we should put some of the worker logic here since it is too verbose compared to other workers (like `root-block-archiver`).

liuchengxu replied:
Well, it's merely about 10 lines, I don't think that's too much, the initialize and update are two function calls. But this change will make the reader easily see what's happening here without looking into the actual worker.

NingLin-P replied:
> But this change will make the reader easily see what's happening here without looking into the actual worker.

I am not sure this is what we need to achieve here as this function already ~300 locs, and these codes are some implemented detail that should keep inside its won crate and it is not always interested by people who come to this function.

Indeed, it may be nice for some newcomers to the codebase to have a look at this function and know what component/worker a node consists of and what they roughly doing. But besides that, perhaps most of the time people who come to this function just want to scroll the mouse away from these details and find the exact component/worker they are interested (within ~300 locs).

But I am not strongly opposed though and will make these changes if you insist.

liuchengxu replied:
It's not a hard requirement, but why not if it can be achieved easily?

No strong opinion either as long as you can make it descriptive and understandable, I can guess what the other tasks are for without looking into them, but not this one.

NingLin-P replied:
The bundle-collector worker logic is outlined in `new_full`.

> the initialization of stored bundles in the last k is implicit on the first imported block notification, which is not intuitive and incorrect even because new bundles can be received before the very first block notification

But this part is unchanged because I found that as the node is just started, it will need a major sync to sync block from other peers, and before that finish, it can only collect bundle from some old block instead the last K block, so it doesn't help much to initialize the bundles on startup.

Also, I found that the `import_notification_stream` may start to notify block import before it actually synced to the tip of the chain (more specifically, 5 blocks before the tip of the chain):

https://github.com/paritytech/substrate/blob/2a636e26e9c200bb77910e4956210bce2bef0e22/client/network/sync/src/lib.rs#L489-L502

and as the transaction in txn pool will be re-validated after a new block is imported, so it may be okay to initialize on the first imported block notification.

liuchengxu replied:
>as the transaction in txn pool will be re-validated after a new block is imported, so it may be okay to initialize on the first imported block notification.

Well, the problem I mentioned before still persists, it's possible to include the duplicated bundle in the very first block if it's not initialized. So we should initialize it only when the node is not in major sync on startup.

NingLin-P replied:
The node will not building block while in major sync as the slot worker will skip propose slot. After the major sync finish, the node still have 5 more blocks waiting to import, while these blocks import the bundle will be initialized and the txn will be re-validated.

But indeed there is a corner case where the major sync just finished and before next block import notification come, the next slot come first and the node just win the challange and build the block without duplicated-bundle check, then created a fork that at least 5 block behide the best fork.

And I think `initialize it only when the node is not in major sync on startup` doesn't do any help to this case.

liuchengxu replied:
> After the major sync finish, the node still have 5 more blocks waiting to import

The node may or may not have 5 more blocks to import, say, you just restart the node when it's already synced to the tip, in which case it has no newer blocks to import before producing the next block.

What we're trying to achieve is essentially to perform the duplicate bundle check to prevent them into the newer blocks, which requires whenever the major sync is false (slot worker will attempt to claim the slot) the bundles in last K should be available.

liuchengxu commented on crates/subspace-bundle-checker/src/lib.rs:
```diff
@@ -0,0 +1,295 @@
+        let bundle_hashs: HashSet<_> = self
+            .client
+            .runtime_api()
+            .extract_bundle_hashs(&BlockId::Hash(block_hash), extrinsics)?
```
We can't collect the bundles stored in a block this way because not every bundle in the block will be stored, there is a chance that some of them may fail. You should retrieve the `BundleStored` events in this block instead.

liuchengxu reviewed (changes requested):
The idea makes sense in general, but the implementation needs a major rewrite.

Given this PR will be changed a lot, I suggest rebasing first before working on it again. And one general comment, before opening up a PR, try rebasing it locally and reorganizing the commits to be more understandable, the commit of fixing the bug in the new code usually don't have to be present in the end, you can squash it into the commit in which the code was initially introduced.

NingLin-P committed:
Introduce the extract_stored_bundle_hashes api to ExecutorApi

It is used to extract the hashes of bundles stored in the block,
which will be used to validate duplicate bundle

Signed-off-by: linning <linningde25@gmail.com>

liuchengxu commented on crates/subspace-transaction-pool/src/bundle_validator.rs:
```diff
@@ -0,0 +1,256 @@
+            .map(|(block_hash, _)| *block_hash)
+    }
+
+    fn collect_bundle_from_block(
```
How about `{extract_}stored_bundles_at` and just returning `HashSet<Hash>` as `block_hash` is the same so that it can be written like this below:

```rust
            let bundles = self.stored_bundles_at(block_hash)?;
            self.bundle_stored_in_last_k.push_front((block_hash, bundles));
```
This is more readable to me.

liuchengxu commented on crates/subspace-transaction-pool/src/bundle_validator.rs:
```diff
@@ -0,0 +1,256 @@
+        &self,
+        at: BlockId<Block>,
+        signed_opaque_bundle: &SignedOpaqueBundle<NumberFor<Block>, Block::Hash, DomainHash>,
+    ) -> sp_blockchain::Result<()>;
```
I suggest using a dedicated error type `BundleError` so that it's more straightforward and readable, non-blocker.

liuchengxu commented on crates/subspace-transaction-pool/src/bundle_validator.rs:
```diff
@@ -0,0 +1,256 @@
+    /// Collect bundles within the new blocks of the best chain, blocks are handled from
+    /// older blcok to newer blcok, an `Err` may return in the middle and left some blocks
+    /// unhandled, these blocks will be handled when processing the next new best block.
+    fn on_import_new_best_block(
```
Hmm, how about just `on_new_best_block` or `on_new_best_block_imported` just to be more accurate?

liuchengxu commented on crates/subspace-transaction-pool/src/bundle_validator.rs:
```diff
@@ -0,0 +1,256 @@
+        }
+    }
+
+    fn update(&self, incoming: VecDeque<(Block::Hash, HashSet<Hash>)>) {
```
```suggestion
    fn update(&self, new_last_k_bundles: VecDeque<(Block::Hash, HashSet<Hash>)>) {
```

liuchengxu commented on crates/subspace-transaction-pool/src/bundle_validator.rs:
```diff
@@ -0,0 +1,256 @@
+        &mut self,
+        mut hash: Block::Hash,
+    ) -> Result<(), sp_blockchain::Error> {
+        assert!(self.bundle_stored_in_last_k.is_empty());
```
An error is usually preferred.

NingLin-P replied:
Error is unnecessary, this is kind of a defense (like the `expect`, and `panic`), the caller to this private function will check `bundle_stored_in_last_k` is empty before calling it.

liuchengxu replied:
Then a message is better to be added to explain why it's okay.

vedhavyas replied:
Why assert though? not sure I understand the reason though. We can just fill up the items and truncate, no ?

NingLin-P replied:
We can not, the `bundle_stored_in_last_k` is storing the bundle hash from the last **consecutive** K blocks of the same fork, if `bundle_stored_in_last_k` is not empty calling this function may cause its state mixup (i.e non-consecutive). 

Also, this function is for initialization and it should only call once when `bundle_stored_in_last_k` is empty. (it will be renamed to `initialize_recent_stored_bundles` as @liuchengxu suggested)

liuchengxu commented on crates/subspace-transaction-pool/src/bundle_validator.rs:
```diff
@@ -0,0 +1,256 @@
+
+/// `BundleStoredInLastK` used to share the bundles of the last K blocks with different thread
+#[derive(Clone)]
+struct BundleStoredInLastK<Block: BlockT> {
```
I don't understand why `BundleCollector` (and this is not a good name) is even needed. Wouldn't just a sharable `BundleStoredInLastK` be simpler?

```rust
struct BundleStoredInLastK {
    client: Arc<Client>,
    bundle_stored_in_last_k: Arc<Mutex<VecDeque<Block::Hash, HashSet<Hash>>>>,
    confirmation_depth_k: usize,
}

impl BundleStoredInLastK {
    pub fn new(client: Arc<Client>) -> Self {
    }

    pub async fn on_new_best_block(&self) {
        // Update `bundle_stored_in_last_k`
    }

    pub fn is_duplicated_bundle(&self, bundle_hash: Hash) -> bool {
        // Return whether `bundle_hash` exists in `bundle_stored_in_last_k`
    }
}

struct BundleValidator {
    client,
    bundle_stored_in_last_k: BundleStoredInLastK
}

impl BundleValidator {
    pub fn new(client: Client, bundle_stored_in_last_k: BundleStoredInLastK) -> Self {
    }
}

impl BundleValidator {
    fn validate_bundle(&self, bundle) {
        // Check the duplicated bundle.
        let bundle_hash = bundle.hash();
        if self.bundle_stored_in_last_k.is_duplicated_bundle(bundle_hash) {
            return Err(BundleError::Duplicated)
        }

        // Check the receipts
    }
}
```

NingLin-P replied:
The `BundleStoredInLastK` you proposed here almost like identical to `BundleCollector`, so except the naming I guess your question here would be: why in current implement there are both `bundle_syncer` and `bundle_stored_in_last_k`
```rust
bundle_syncer: BundleStoredInLastK<Block>, // namely `Arc<Mutex<VecDeque<Block::Hash, HashSet<Hash>>>>`
bundle_stored_in_last_k: VecDeque<(Block::Hash, HashSet<Hash>)>,
```
instead of just `bundle_syncer` and make `BundleCollector` shared between thread?

The reason is while maintaining `bundle_stored_in_last_k` in `on_import_new_best_block` we need to access this filed quite many time and in between there may include some database reading, we do not want to doing that while holding a `Mutex` as it will blocking txn validation.

We can also twist the logic in `on_import_new_best_block` to make it avoid database reading while holding the `Mutex`, but that will hurt its readability and maintainability which is more important I think.

liuchengxu replied:
I'm unsure we should take into account the database operations here because a) the transactions are validated concurrently, it has no influence on the other transactions except `submit_bundle`, and I think we probably ought to even _block_ here in order to have a complete check on the bundles. b) I don't see the database operations here are heavy/special, every transaction validation involves the database and runtime API call.

NingLin-P replied:
Well, whether database operations are heavy/special or not, it is good to remove them completely from this check though.

The code you propose here does reduce a few locs by combining the bundle set maintenance and sharing into one, but I prefer the current implementation where the logic of maintenance and sharing are separated because it improves the readability and maintainability of both maintenance and sharing.

liuchengxu replied:
>it improves the readability and maintainability of both maintenance and sharing.

I guess that's the problem :P, my opinion is on the opposite, I don't see the evidence of performance benefits and it makes the implementation over complicated unnecessarily, which is harmful to the readability and maintainability.

NingLin-P replied:
Seems we got an opposing view :)

NingLin-P replied:
The maintenance is about reading/updating the bundle set, when combining maintenance and sharing into one, the bundle set is behind a `Mutex`. 

And when comes to `Mutex`, people reading or changing the code need to think about how long the lock gonna hold, whether there will be a deadlock, will the lock conflict be frequent, and how to ensure the update to bundle set is atomic, etc. And this kind of burden is why I think it hurt readability and maintainability.

When maintenance and sharing are separated, this burden is minimized because the sharing part is scoped into a few lines of very simple code. And the maintenance part just needs to focus on how to reading/updating the bundle set when a new best block is imported. And this is the reason I think it improves readability and maintainability.

liuchengxu replied:
It's kind of over-engineering from what I see, there are a lot of Mutex usages in our codebase, but I haven't seen a similar pattern.

NingLin-P replied:
Well, in my point of view, both the current implementation and your proposal work, the discussion here is just about some subtle personal coding style and preference.

And If we have to have a discussion like this to get a consensus, then the abovementioned is what backs my opinion, but that doesn't mean every time I have to spend a lot of time thinking to get something like this implemented because it is just natural to me.

And I do believe it does not bring complexity and has the abovementioned pros, and why not why have to change to another way? If it is indeed important and worth the time then let's discuss more the pros & cons of different preferences to get a consensus.

liuchengxu replied:
My suggestion was from a standpoint that one general principle is to prefer easy simple understandable code and try to keep consistency with the existing code if possible. The pros you mentioned were not convincing to me for this use case, but I'm not a dictator here. It's absolutely fair to have a personal coding preference, but I can't silently approve the code which looks alien to the existing ones without a decent reason. Could you briefly make a conclusion of our discussion (I guess the main point of your arguments is Mutex) here and post it on slack to get more people's opinions?

nazar-pc replied:
TL;DR: I'm always in favor of misuse-resistant code, see my understanding and proposal below.

I agree `bundle_stored_in_last_k` is not pretty, but I'm not sure removing it will be a good idea, locks will indeed have to be held for quite some time.

I'm not sure if that is a problem though, it depends on whether there is a concurrent thing happening that will be negatively affected by this. Also the way it is implemented right now is NOT atomic in logical sense because you can plausibly have two `BundleCollector`s that each have their own `bundle_stored_in_last_k` and will basically step on each other toes overriding contents of `bundle_syncer` with conflicting values because their internal logic is NOT actually based on `bundle_syncer`'s contents.

Even if we need to have two of these, I think there should be be a data structure that holds both fields behind mutexes in there and ensures invariants are preserved and we never get a diverging view like I explained in above paragraph. This can be achieved by doing updates in a callback or returning something similar to mutex guard that'll call `bundle_syncer.update()` internally on drop.

That way most of the time quick lock will be taken on `bundle_syncer` and sometimes longer lock will be taken on `bundle_stored_in_last_k`, so performance generally will be identical to current implementation with additional security guarantees. As to deadlocks, you can do `bundle_stored_in_last_k.try_lock()` if you don't expect it to every be accessed concurrently and return an error if it is already held to signal a bug in business logic.

NingLin-P replied:
> you can plausibly have two BundleCollectors that each have their own bundle_stored_in_last_k and will basically step on each other toes overriding contents of bundle_syncer with conflicting values

Currently, there should only be one `BundleCollector` when the node is running. And because `BundleCollector` didn't `#[derive(Clone)]` if there is more than one `BundleCollector` somehow then they are created by `BundleCollector::new` thus they will not share the same `bundle_syncer`.

We can also make the code more robust by `impl Clone for BundleCollector` and ensure the cloned `BundleCollector` will have a new and empty state though, which prevents people from accidentally `#[derive(Clone)]` for `BundleCollector` get the abovementioned problem.

nazar-pc replied:
It is ideal when the wrong code doesn't compile. My suggestion is to make it so that that particular issue is non-issue. Manually implementing `Clone` has no impact on ability to use derive later.

NingLin-P replied:
Manually implementing `Clone` will make `#[derive(Clone)]` fail to compile due to `conflicting implementations of trait Clone`. And unless someone intentionally changes the code to make two `BundleCollector` sharing the same `bundle_syncer` I think it will be okay.

nazar-pc replied:
It fixes nothing about the API concern I have about `.update()` that should be called with the whole contents.

What do you think about wrapper data structure that encapsulates both though? To the outside it'll look like one data structure, but it'll be more efficient.

NingLin-P replied:
Okay, I think I got your concern about the `.update()` and the diverging view. I had written a little prototype of your proposal, and it looks good to me as it also fulfills what I need, will push it here after some polish and see how it goes.

liuchengxu commented on crates/subspace-service/src/lib.rs:
```diff
@@ -248,6 +253,13 @@ where
+        let confirmation_depth_k = get_chain_constants(client.as_ref())
+            .expect("Must always be able to get chain constants")
+            .confirmation_depth_k();
+        BundleCollector::new(client.clone(), confirmation_depth_k as usize)
```
Why not invoke `get_chain_constants()` inside BundleCollector in order to simplify the construction here?

rahulksnv commented on crates/subspace-transaction-pool/src/bundle_validator.rs:
```diff
@@ -0,0 +1,256 @@
+#[derive(Clone)]
+struct BundleStoredInLastK<Block: BlockT> {
+    #[allow(clippy::type_complexity)]
+    bundle_stored_in_last_k: Arc<Mutex<VecDeque<(Block::Hash, HashSet<Hash>)>>>,
```
If multiple concurrent readers(validators) are possible, RwLock would be a better fit here

rahulksnv commented on crates/subspace-transaction-pool/src/bundle_validator.rs:
```diff
@@ -0,0 +1,256 @@
+        // The hash used here must be the same as what is maintaining in `bundle_stored_in_last_k`,
+        // namely the hash of `SignedOpaqueBundle`
+        let incoming_bundle = signed_opaque_bundle.hash();
+        let block_bundles = self.bundle_syncer.bundle_stored_in_last_k.lock();
```
Since BundleStoredInLastK has an update() method, it would also be preferable to have a similar `is_bundle_present()` (or some such name). That way callers don't need to know about internal locking mechanism, and lock can be dropped sooner in the called function. Think @liuchengxu suggested something similar. 

If we go with the current way, please drop the lock sooner (using `std::mem::drop()`), after L238.

NingLin-P replied:
Right, having a dedicated method for this check will be helpful.

nazar-pc replied:
It doesn't seem that this lock needs to be held for this long, maybe wrap with `for` in `{}` to ensure it is dropped sooner?

rahulksnv commented on crates/subspace-transaction-pool/src/bundle_validator.rs:
```diff
@@ -0,0 +1,256 @@
+    }
+
+    fn update(&self, incoming: VecDeque<(Block::Hash, HashSet<Hash>)>) {
+        *self.bundle_stored_in_last_k.lock() = incoming;
```
Pls consider doing `a std::mem::swap()` with an empty `VecDequeue<>` under lock, so that the actual free/heap operations happen outside the lock

nazar-pc replied:
1) It is not even a micro, it is a nano-optimization
2) Are you sure that'll be the case? I didn't verify this particular case, but in most other cases Rust actually drops things in reverse order, so it might end up being the same, but will hurt readability quite a bit.

rahulksnv replied:
yeah it only helps if the the VecDequeue being freed is big enough. With k = 100, and lot of transactions per block, this would be helpful.

And I'm sure. Because the assignment would trigger the dtor of the previous VecDequeue, while lock is  still held

vedhavyas commented on crates/subspace-transaction-pool/src/bundle_validator.rs:
```diff
@@ -0,0 +1,256 @@
+        assert!(self.bundle_stored_in_last_k.is_empty());
+        // `block_hashes` sorted from older block to newer block
+        let mut block_hashes = VecDeque::new();
+        for _ in 0..self.confirm_depth_k {
```
this hurts readability for me. This look loops forwards from 0 and ignores index and inner logic seems to go opposite. Why not make outer also have a consistent flow with inner loop?

NingLin-P replied:
`for _ in 0..x` is for counting only.

vedhavyas commented on crates/subspace-transaction-pool/src/bundle_validator.rs:
```diff
@@ -0,0 +1,256 @@
+            match self.client.header(hash)? {
+                Some(header) if !header.number().is_zero() => hash = *header.parent_hash(),
+                // Other cases means the chain currently do not have more block
+                _ => break,
```
I think we want to return error here. Do we expect to have this header missing?

NingLin-P replied:
> Do we expect to have this header missing?

Only for the `header.number().is_zero()` case I think.

vedhavyas commented on crates/subspace-transaction-pool/src/bundle_validator.rs:
```diff
@@ -0,0 +1,256 @@
+}
+
+#[derive(Clone)]
+pub struct SkipBundleValidation;
```
Cant we just use empty tuple and short circuit validation in default cases?

liuchengxu replied:
An empty tuple was used initially, but it looks less straightforward on the call site, see https://github.com/subspace/subspace/blob/4033d3706b/domains/service/src/system_domain.rs#L187-L193, thus I suggested using something more explicit and descriptive. Feel free to suggest a better alternative.

NingLin-P replied:
`()` is used originally, @liuchengxu suggested `SkipBundleValidation` as it is more informative I also agree with that.

vedhavyas replied:
ahh I see. We always seem to use empty tuple for default impls. Explicit is fine as well

liuchengxu replied:
Yeah, we use empty tuples a lot while constructing the runtime and feel comfortable there, but using `()` suddenly in such a place is somewhat not preferable.

vedhavyas commented on crates/subspace-transaction-pool/src/bundle_validator.rs:
```diff
@@ -0,0 +1,256 @@
+            }
+        }
+
+        let block_number =
```
better to call to it `best_primary_number` since this anyway assumes receipts are always constructed with number beloe best

vedhavyas commented on domains/service/src/system_domain.rs:
```diff
@@ -183,8 +185,13 @@ where
+        &task_manager,
+        client.clone(),
+        proof_verifier,
+        SkipBundleValidation,
```
Some comment on why bundle verification is skipped here would be useful

vedhavyas reviewed (commented):
Makes sense. Instead of maintaining copies, I would prefer using a RWMutex since there is essentially only one writer.

NingLin-P committed:
Introduce the bundle_validator module to subspace-transaction-pool

It will be used to validate receipt and duplicated bundle within the txn pool.

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Use BundleValidator to validate bundle within the txn pool

Signed-off-by: linning <linningde25@gmail.com>

liuchengxu commented on crates/subspace-transaction-pool/src/bundle_validator.rs:
```diff
@@ -0,0 +1,327 @@
+                return true;
+            }
+        }
+        false
```
This could be more concise.

```Rust
block_bundles.iter().any(|(_, bundle_hashes)| bundle_hashes.contains(&hash))
```

liuchengxu commented on crates/subspace-node/src/bin/subspace-node.rs:
```diff
@@ -199,7 +199,7 @@ fn main() -> Result<(), Error> {
                     import_queue,
                     task_manager,
-                    other: (_block_import, subspace_link, _telemetry),
+                    other: (_block_import, subspace_link, _telemetry, _bundle_collector),
```
```suggestion
                    other: (_block_import, subspace_link, _telemetry, _bundle_validator),
```

liuchengxu commented on crates/subspace-transaction-pool/src/bundle_validator.rs:
```diff
@@ -0,0 +1,327 @@
+                )))
+            }
+        };
+        let pre_best_hash = BundleStoredInLastK::<Block>::best_hash(&bundles);
```
```suggestion
        let pre_best_hash = Self::best_hash(&bundles);
```

liuchengxu commented on crates/subspace-transaction-pool/src/bundle_validator.rs:
```diff
@@ -0,0 +1,327 @@
+        };
+        let pre_best_hash = BundleStoredInLastK::<Block>::best_hash(&bundles);
+        let res = update_fn(&mut bundles);
+        if pre_best_hash != BundleStoredInLastK::<Block>::best_hash(&bundles) {
```
```suggestion
        if pre_best_hash != Self::best_hash(&bundles) {
```

liuchengxu reviewed (commented):
Mostly make sense to me. It'd be great to add a test for it, but I understand it's not easy given the existing testing framework.

NingLin-P committed:
Apply review suggestions

Signed-off-by: linning <linningde25@gmail.com>

liuchengxu reviewed (approved):
Overall reasonable to me except that I'm not sure about the complexity of some implementation details.

NingLin-P committed:
Merge branch 'main' into prevent-duplicated-bundles
